### DISCUSSION TITLE
dev: webpack bundle sizes

### DISCUSSION THREAD

**Discussion by danielweck** [edrlab/thorium-reader#1758] [Developer corner]
Perhaps there is some room for improvement ...
(removing Moment locales would decrease the bundle size, for example)

<img width="515" alt="Screenshot 2020-10-15 at 17 30 07" src="https://user-images.githubusercontent.com/625201/96159081-259cf680-0f0c-11eb-966c-792a5c91a5d6.png">

<img width="1792" alt="Screenshot 2020-10-15 at 17 26 08" src="https://user-images.githubusercontent.com/625201/96158759-be7f4200-0f0b-11eb-8549-299cf91e4a67.png">

<img width="1792" alt="Screenshot 2020-10-15 at 17 24 44" src="https://user-images.githubusercontent.com/625201/96158684-ac050880-0f0b-11eb-9df7-4c86daae13b9.png">

<img width="1792" alt="Screenshot 2020-10-15 at 17 25 28" src="https://user-images.githubusercontent.com/625201/96158713-b2938000-0f0b-11eb-8a68-f067738895d4.png">

**Comment by danielweck** [comment]
PS: note that bundle compression/optimization is enabled only to remove whitespaces, not to mangle class/function names, and not to perform any other smart processing of the Javascript AST.

See:

https://github.com/edrlab/thorium-reader/blob/bee2de14c0f3b3536ff72bef9b0c6c08fe461727/webpack.config.main.js#L203-L225

https://github.com/edrlab/thorium-reader/blob/bee2de14c0f3b3536ff72bef9b0c6c08fe461727/webpack.config.preload.js#L28-L48

https://github.com/edrlab/thorium-reader/blob/bee2de14c0f3b3536ff72bef9b0c6c08fe461727/webpack.config.renderer-library.js#L274-L294

https://github.com/edrlab/thorium-reader/blob/bee2de14c0f3b3536ff72bef9b0c6c08fe461727/webpack.config.renderer-pdf.js#L270-L290

https://github.com/edrlab/thorium-reader/blob/bee2de14c0f3b3536ff72bef9b0c6c08fe461727/webpack.config.renderer-reader.js#L262-L282

**Comment by danielweck** [comment]
This commit introduces `npm run bundle-stats` to generate the bundle contents images:
https://github.com/edrlab/thorium-reader/commit/51c1bafa59b466e0f4ecaa35af7b69d1ba9e9bdc

**Comment by MattiJarvinen** [comment]
I'd advice switching from momentjs to date-fns, you can then drop out all momentjs localizations, if you need all the localizations then there's nothing you can do there.

There seems to be quite a lot of lodash... depending on your browser targets most of the lodash stuff is easy to accomplish with modern JS or just tweak your imports to tree shakeable. You might find some help on that here https://www.azavea.com/blog/2019/03/07/lessons-on-tree-shaking-lodash/. Especially the tree shaking part could help to reduce that quite a bit. 

> Tree-shaking is syntax-sensitive. To enable tree-shaking in lodash, you have to import functions using syntax like import foo from 'lodash/foo'. import { foo } from 'lodash' will not tree-shake, nor will, obviously, import _ from 'lodash'. Support for this syntax was implemented in Lodash v4.

Anyhow for a packaged / desktop application this isn't so much of an issue, but a good thing to keep your eyes on.

**Comment by danielweck** [comment]
Thank you for your input. Some of the heavy dependencies are transitive, so we might need to dig deeper to find the culprits.

**Comment by danielweck** [comment]
The Javascript bundles generated by WebPack for the "reader" and "library"  renderer Electron processes include code that is shared with / common to the "main" process bundle. Therefore, there is an opportunity to factor out some dependencies by placing them into `node_modules` (i.e. still located inside the Electron ASAR, but inside an externalised `node_modules` folder within the ASAR virtual filesystem). This is technically doable because at the moment the "reader" and "library"  renderer Electron processes are "Node enabled", which means that the NodeJS dependency resolver can be used instead of relying solely on WebPack's own bundler / resolver. A simple way to implement this would be to include additional NPM dependencies in WebPack's `externals` directive. The net gain would be x-amount of bytes in the main and renderer processes JS bundles ... to be determined empirically (i.e. by implementing the WebPack change, and comparing the size of generated bundles).
Note that preferably ("eventually"?) all renderer Electron processes should have NodeJS disabled (aka "node integration"), and rely purely on postMessage/WebSocket/HTTP etc. rather than IPC Electron-specific APIs. I'm not sure how extensive the refactoring would need to be ... probably quite disruptive and prone to introduce regression bugs.

**Comment by danielweck** [comment]
PS: when "node integration" is disabled in renderer processes, standard OpenWebPlatform APIs can be used (such as postMessage/WebSocket/HTTP), but there is probably a better alternative: `preload`. The reader and library BrowserWindows could implement a "communication bridge" in a preload script that would have high NodeJS privileges, including the capacity to resolve NodeJS packages and use Electron/Node APIs (such as the IPC layer).
